docs: add test examples with undici and fetch (#4300)
fastify/fastify: docs/Guides/Testing.md

const { Client } = require('undici')
const buildFastify = require('./app')

tap.test('should work with undici', async t => {
  t.plan(5)

  const fastify = buildFastify()

  t.teardown(() => fastify.close())

  await fastify.listen()

  const client = new Client(
    'http://localhost:' +
    fastify.server.address().port
  )

  const response = await client.request({ method: 'GET', path: '/' })

  t.equal(await response.body.text(), 'Hello from fastify-slow-down!')
  t.equal(response.statusCode, 200)
})
```
**test-fetch.js** (testing with
[`fetch`](https://nodejs.org/docs/latest-v18.x/api/globals.html#fetch))
> **_NOTE:_** `fetch` is available in Node 18
```js
const tap = require('tap')
const buildFastify = require('./app')

tap.test('should work with fetch', async t => {
  t.plan(5)

  const fastify = buildFastify()

  t.teardown(() => fastify.close())

  await fastify.listen()
  
  const response = await fetch(
    'http://localhost:' +
    fastify.server.address().port
  )

  t.equal(response.status, 200)
  t.equal(
    response.headers.get('content-type'),
    'application/json; charset=utf-8'
  )
  t.has(await response.json(), { hello: 'world' })
})

> ```

Reviewer: Why duplicate the example?

Author: @jsumners is not a duplicate, the first one is an example using undici and the second example using fetch from Node 18

Reviewer: They do exactly the same thing. Thus, they are duplicate examples. Are we trying to say fetch is so different that someone using it can't understand the basic client example? It doesn't look so to me from the provided examples.

Reviewer: That is not clear in the documentation as it is presented. It just seems like a second snippet that is very slightly different for no discernible reason.

Author: I will add between the `test-undici` and `test-fetch` a section `Another example`, the readers will understand that is the same approach but is using another package

Uses **app.js** from the previous example.

**test-undici.js** (testing with [`undici`](https://www.npmjs.com/package/undici))
```js
const tap = require('tap')
const { Client } = require('undici')
const buildFastify = require('./app')

tap.test('should work with undici', async t => {
  t.plan(5)

  const fastify = buildFastify()

  t.teardown(() => fastify.close())

  await fastify.listen()

  const client = new Client(
    'http://localhost:' + fastify.server.address().port, {
      keepAliveTimeout: 10,
      keepAliveMaxTimeout: 10
    } 

>   )

Reviewer 2: ```suggestion
  )
  t.teardown(() => client.close())
```

Author: Thanks!

Uses **app.js** from the previous example.

**test-undici.js** (testing with [`undici`](https://www.npmjs.com/package/undici))
```js
const tap = require('tap')
const { Client } = require('undici')
const buildFastify = require('./app')

tap.test('should work with undici', async t => {
  t.plan(2)

  const fastify = buildFastify()

  await fastify.listen()

   const client = new Client(
    'http://localhost:' + fastify.server.address().port, {
      keepAliveTimeout: 10,
      keepAliveMaxTimeout: 10
    } 
  )

  t.teardown(() => {
    fastify.close()
    client.close()
  })

  const response = await client.request({ method: 'GET', path: '/' })

>   t.equal(await response.body.text(), 'Hello from fastify-slow-down!')

Reviewer 3: Shouldn't response text be `{"hello":"world"}` ?

Reviewer 4: ```suggestion
  t.equal(await response.body.text(), '{"hello":"world"}')
```

Uses **app.js** from the previous example.

**test-listen.js** (testing with [`undici`](https://www.npmjs.com/package/undici))
```js
const tap = require('tap')
const { Client } = require('undici')
const buildFastify = require('./app')

tap.test('should work with undici', async t => {
  t.plan(2)

  const fastify = buildFastify()

  await fastify.listen()

   const client = new Client(
    'http://localhost:' + fastify.server.address().port, {
      keepAliveTimeout: 10,

>       keepAliveMaxTimeout: 10

Reviewer 5: it may be worth pointing out why these are needed. I think there's a section in undici's docs about this right?

Author: Yes, here is the `undici` documentation:  https://undici.nodejs.org/#/docs/api/Client

Uses **app.js** from the previous example.

**test-listen.js** (testing with [`undici`](https://www.npmjs.com/package/undici))
```js
const tap = require('tap')
const { Client } = require('undici')
const buildFastify = require('./app')

tap.test('should work with undici', async t => {
  t.plan(2)

  const fastify = buildFastify()

  await fastify.listen()

   const client = new Client(
    'http://localhost:' + fastify.server.address().port, {
      keepAliveTimeout: 10,
      keepAliveMaxTimeout: 10
    } 
  )

  t.teardown(() => {
    fastify.close()
    client.close()
  })

  const response = await client.request({ method: 'GET', path: '/' })

  t.equal(await response.body.text(), '{"hello":"world"}')
  t.equal(response.statusCode, 200)
})
```

> 
> #### Another Example:
> 
> This example is a copy of the previous one, this time using `fetch` 
> to make the request. 
> 
> > **_NOTE:_** `fetch` is available in Node 18
> 
> Uses **app.js** from the previous example.
> 
> **test-listen.js** (testing with
> [`fetch`](https://nodejs.org/docs/latest-v18.x/api/globals.html#fetch))

Reviewer: I don't know why this suggestion didn't get added in my last review. It was supposed to be. GitHub just decided to ignore it completely 😕 

```suggestion

Alternatively, starting with Node.js 18, [`fetch`](https://nodejs.org/docs/latest-v18.x/api/globals.html#fetch) may be used without requiring any extra dependencies:

**test-listen.js**
```
